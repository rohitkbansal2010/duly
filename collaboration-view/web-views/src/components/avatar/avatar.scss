@import "~@styles/colors.scss";
@import "~@styles/typography.scss";

.avatar {
  display: flex !important;
  flex-shrink: 0 !important;
  border-radius: 50%;

  img {
    z-index: 2;
    width: 100%;
    height: 100%;
  }

  &Fallback {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    z-index: 1 !important;
  }

  &Lettered {
    position: absolute;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    display: flex !important;
    align-items: center !important;
    justify-content: center !important;
    width: 100% !important;
    height: 100% !important;
    color: $white;
    font-weight: 600;
    line-height: 100%;
    border-radius: 50%;
    @include nexa-extra-bold-highlight;

    &MedicalDoctor,
    &RegularEmployee {
      margin: 0 auto !important;
    }

    &MedicalDoctor {
      background-color: $dupage-dark-blue;
    }

    &RegularEmployee {
      background-color: $dupage-magenta;
    }

    &Transparent {
      background-color: transparent;
    }
  }
}

.borderAvatar {
  border-color: $white;
  border-style: solid inset solid solid;

  &RegularEmployee,
  &MedicalDoctor,
  &NonEmployee {
    @extend .borderAvatar;
    outline-style: solid;
  }

  &NonEmployee {
    outline-color: $white;
    box-shadow: 0 0.125rem 0.25rem -0.0625rem rgb(102 110 125 / 6%),
      0 0.25rem 0.375rem -0.0625rem rgb(102 110 125 / 10%), 0 0 0.0625rem 0 rgb(102 110 125 / 6%);
  }

  &RegularEmployee {
    outline-color: $dupage-violet;
  }

  &MedicalDoctor {
    outline-color: $dupage-light-blue;
  }
}
