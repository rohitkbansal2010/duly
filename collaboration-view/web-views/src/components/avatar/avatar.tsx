import { useState } from 'react';
import { Image } from 'react-bootstrap';

import { AvatarByRole } from '@enums';
import { nonEmployeeIcon } from '@icons';

import styles from './avatar.scss';

type AvatarProps = {
  src?: string;
  alt?: string;
  width?: number;
  role?: AvatarByRole;
  hasBorder?: boolean;
}

const INITIALS_LENGTH = 2;

const getInitials = (nameString: string) =>
  nameString
    .split(' ')
    .slice(0, INITIALS_LENGTH)
    .map(name =>
      name[0])
    .join('')
    .toUpperCase();

export const Avatar = ({
  src,
  alt = 'avatar',
  width = 1.5,
  role = AvatarByRole.REGULAR_EMPLOYEE,
  hasBorder = false,
}: AvatarProps) => {
  const [ error, setError ] = useState<boolean>(false);
  const [ loaded, setLoaded ] = useState<boolean>(false);
  const isImageShown = src && !error;

  const avatarClassName = []
    .concat(hasBorder ? styles[`borderAvatar${role}`] : '')
    .concat(styles.avatar)
    .join(' ');

  const letteredClassName = []
    .concat(styles.avatarLettered)
    .concat(loaded ? styles.avatarLetteredTransparent : styles[`avatarLettered${role}`])
    .join(' ');

  return (
    <div
      className={avatarClassName}
      style={{
        width: `${width}rem`,
        height: `${width}rem`,
        position: 'relative',
        fontSize: `${width / 3}rem`,
        borderWidth: `${width * 0.02}rem`,
        outlineWidth: `${width * 0.03}rem`,
      }}
    >
      {isImageShown && (
        <Image
          src={src}
          alt={alt}
          title={alt}
          width="100%"
          height="100%"
          roundedCircle
          onLoad={() =>
            setLoaded(true)}
          onError={() =>
            setError(true)}
        />
      )}
      {(!loaded || error) && (
        role === AvatarByRole.NON_EMPLOYEE
          ? (
            <Image
              className={styles.avatarFallback}
              src={nonEmployeeIcon}
              alt={alt}
              title={alt}
              roundedCircle
            />
          )
          : (
            <div className={letteredClassName}>
              {!isImageShown && getInitials(alt)}
            </div>
          )
      )}
    </div>
  );
};
