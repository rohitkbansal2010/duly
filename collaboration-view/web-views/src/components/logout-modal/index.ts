export { LogoutModal } from './logout-modal';
