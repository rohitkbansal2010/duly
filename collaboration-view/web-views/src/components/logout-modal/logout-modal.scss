@import "~@styles/colors.scss";
@import "~@styles/typography.scss";

.logoutModal {
  &Header {
    justify-content: center !important;
    margin-bottom: 4.25rem;
    padding: 0 !important;
    border: 0 !important;
    border-radius: 0 !important;

    & img {
      width: 12.75rem;
      height: 5.94rem;
    }
  }

  &Body {
    padding: 0 !important;

    &PractitionerLogo {
      display: flex;
      align-items: center;
      justify-content: center;
      width: 12.4rem;
      height: 12.4rem;
      margin: 0 auto;
      margin-bottom: 1.31rem;
      background: $white;
      border-radius: 50%;
    }

    &PractitionerHumanName {
      margin-bottom: 1.37rem;
      color: $white;
      font-weight: 800;
      font-size: 3.75rem;
      line-height: 5.63rem;
      @include nexa-heavy-highlight;
    }

    &Question {
      margin-bottom: 2rem;
      color: $white;
      font-weight: 400;
      font-size: 1.5rem;
      line-height: 1.88rem;
      @include nexa-regular-highlight;
    }

    &ButtonDarkBlue {
      background-color: $dupage-dark-blue;
      border: 0.06rem solid $dupage-dark-blue;
    }

    &ButtonViolet {
      background-color: $dupage-violet;
      border: 0.06rem solid $dupage-violet;
    }

    &ButtonLogout,
    &ButtonCancel {
      width: 100%;
      max-width: 13.75rem;
      padding: 1.63rem 0 !important;
      color: $white;
      font-size: 1.25rem !important;
      line-height: 1.5rem !important;
      border-radius: 0.5rem !important;
      @include nexa-extra-bold-highlight;
    }

    &ButtonLogout {
      margin-right: 0.63rem;
    }

    &ButtonCancel {
      background-color: transparent;
    }
  }
}
