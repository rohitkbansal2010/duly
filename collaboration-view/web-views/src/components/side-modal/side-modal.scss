@import "~@styles/colors.scss";
@import "~@styles/typography.scss";

.sideModal {
  &Backgrop {
    opacity: 0 !important;
  }

  &Dialog {
    max-width: 31.13rem !important;
    height: 100%;
    margin: 0 !important;
    margin-left: calc(100% - 31.13rem) !important;
  }

  &Content {
    height: 100%;
    padding: 3.88rem 0.69rem 0 2.13rem;
    border: 0 !important;
    border-radius: 0 0 0 4.13rem !important;
  }

  &ContentBackgroundMagenta {
    background-color: $dupage-magenta !important;
  }

  &ContentBackgroundBlue {
    background-color: $dupage-light-blue !important;
  }
}
