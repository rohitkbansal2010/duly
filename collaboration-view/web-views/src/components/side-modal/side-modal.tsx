import { PropsWithChildren } from 'react';
import { Modal } from 'react-bootstrap';

import styles from './side-modal.scss';

type SideModalProps = {
  show: boolean;
  onHide: () => void;
  backgroundContentColor?: string;
};

export const SideModal = ({
  show,
  onHide,
  backgroundContentColor = 'Blue',
  children,
}: PropsWithChildren<SideModalProps>) =>
  (
    <Modal
      animation={false}
      show={show}
      onHide={onHide}
      className={styles.sideModal}
      backdropClassName={styles.sideModalBackgrop}
      dialogClassName={styles.sideModalDialog}
      contentClassName={`${styles.sideModalContent} ${styles[`sideModalContentBackground${backgroundContentColor}`]}`}
    >
      {children}
    </Modal>
  );
