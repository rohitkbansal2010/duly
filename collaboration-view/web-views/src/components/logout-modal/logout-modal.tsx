import { Modal } from 'react-bootstrap';
import { useSelector } from 'react-redux';

import { Avatar } from '@components/avatar';
import { SideModal } from '@components/side-modal';
import { AvatarByRole } from '@enums';
import { useAzureADAuth, useLogoutModal } from '@hooks';
import { dulyLargeWhiteDarkBlueIcon, dulyLargeWhiteVioletIcon } from '@icons';
import { RootState } from '@redux/reducers';
import { getSrcAvatar, showFirstPrefixItem } from '@utils';

import styles from './logout-modal.scss';

export const LogoutModal = () => {
  const { logoutAzureADRedirect } = useAzureADAuth();

  const { isLogoutModalShown, onHideLogoutModal } = useLogoutModal();
  
  const userData = useSelector(({ USER }: RootState) =>
    USER.userData);
  const userRole = useSelector(({ USER }: RootState) =>
    USER.userRole);

  const isRegularEmployee = userRole === AvatarByRole.REGULAR_EMPLOYEE;
  const backgroundContentColor = isRegularEmployee ? 'Magenta' : 'Blue';
  const styleButton = isRegularEmployee ? 'Violet' : 'DarkBlue';

  return (
    <SideModal
      show={isLogoutModalShown}
      onHide={onHideLogoutModal}
      backgroundContentColor={backgroundContentColor}
    >
      <Modal.Header className={styles.logoutModalHeader}>
        <img
          src={isRegularEmployee ? dulyLargeWhiteVioletIcon : dulyLargeWhiteDarkBlueIcon}
          alt="Large duly icon"
        />
      </Modal.Header>
      <Modal.Body className={styles.logoutModalBody}>
        <div className={styles.logoutModalBodyPractitionerLogo}>
          <Avatar
            width={12.31}
            src={getSrcAvatar(userData?.photo)}
            alt={`${showFirstPrefixItem(userData?.humanName.givenNames)}${userData?.humanName.familyName}`.trim()}
            role={userRole as AvatarByRole}
            hasBorder
          />
        </div>
        <div className={styles.logoutModalBodyPractitionerHumanName} data-testid="human-name">
          {userData?.humanName.prefixes && userData?.humanName.prefixes.length
            ? `${showFirstPrefixItem(userData?.humanName.prefixes)}${userData?.humanName.familyName}`.trim()
            : userData?.humanName.givenNames.join(' ')
          }
        </div>
        <div className={styles.logoutModalBodyQuestion}>
          Are you sure you want to log out?
        </div>
        <div>
          <button
            onClick={() =>{
              sessionStorage.removeItem('metric');
              logoutAzureADRedirect();}}
            className={
              `${styles.logoutModalBodyButtonLogout} ${styles[`logoutModalBodyButton${styleButton}`]}`
            }
            data-testid="logout-button"
          >
            Logout
          </button>
          <button
            onClick={onHideLogoutModal}
            className={
              `${styles.logoutModalBodyButtonCancel} ${styles[`logoutModalBodyButton${styleButton}`]}`
            }
            data-testid="cancel-button"
          >
            Cancel
          </button>
        </div>
      </Modal.Body>
    </SideModal>
  );
};
