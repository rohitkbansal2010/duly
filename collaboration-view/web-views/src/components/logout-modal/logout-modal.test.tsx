import 'jest-localstorage-mock';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Modal } from 'react-bootstrap';
import { useSelector } from 'react-redux';

import { Avatar } from '@components/avatar';
import { SideModal } from '@components/side-modal';
import { AvatarByRole } from '@enums';
import { dulyLargeWhiteDarkBlueIcon, dulyLargeWhiteVioletIcon } from '@icons';
import { getSrcAvatar, showFirstPrefixItem } from '@utils';

import { LogoutModal } from './logout-modal';

const mockUserData = {
  userData: {
    humanName: {
      familyName: 'Morris',
      givenNames: [ 'Jane' ],
      prefixes: [ 'Ms', 'Dr' ],
    },
    photo: {
      contentType: 'image/x-png',
      title: 'Photo',
      size: 0,
      data: 'data',
    },
  },
  userRole: AvatarByRole.REGULAR_EMPLOYEE,
};


jest.mock('react-redux', () =>
  ({
    useSelector: jest.fn((selector: (state: unknown) => unknown) => {
      const state = { USER: mockUserData };
      return selector(state);
    }),
  }));

const mockLogoutAzureADRedirect = jest.fn();
const mockOnHideLogoutModal = jest.fn();
jest.mock('@hooks', () =>
  ({
    useAzureADAuth: () =>
      ({ logoutAzureADRedirect: mockLogoutAzureADRedirect }),
    useLogoutModal: () =>
      ({
        isLogoutModalShown: true,
        onHideLogoutModal: mockOnHideLogoutModal,
      }),
  }));

jest.mock('@utils', () =>
  ({
    getSrcAvatar: jest.fn(),
    showFirstPrefixItem: jest.fn(),
  }));

let wrapper: ShallowWrapper;
beforeEach(() => {
  (getSrcAvatar as jest.Mock).mockImplementation(() =>
    'mockSrc');
  (showFirstPrefixItem as jest.Mock).mockImplementation(() =>
    'mockFirstPrefixItem');

  wrapper = shallow(<LogoutModal />);
});

describe('LogoutModal', () => {
  it('should match snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call logoutAzureADRedirect on logout button click', () => {
    const logoutButton = wrapper.find({ 'data-testid': 'logout-button' });
    logoutButton.simulate('click');

    expect(mockLogoutAzureADRedirect).toHaveBeenCalledTimes(1);
  });

  it('should call onHideLogoutModal on cancel button click', () => {
    const cancelButton = wrapper.find({ 'data-testid': 'cancel-button' });
    cancelButton.simulate('click');

    expect(mockOnHideLogoutModal).toHaveBeenCalledTimes(1);
  });

  it('should call showFirstPrefixItem twice with correct params', () => {
    const { givenNames, prefixes } = mockUserData.userData.humanName;

    expect(showFirstPrefixItem).toHaveBeenNthCalledWith(1, givenNames);
    expect(showFirstPrefixItem).toHaveBeenNthCalledWith(2, prefixes);
  });

  it('should call getSrcAvatar with correct params', () => {
    const { photo } = mockUserData.userData;

    expect(getSrcAvatar).toHaveBeenCalledWith(photo);
  });

  it('should render SideModal with correct props', () => {
    const sideModal = wrapper.find(SideModal);
    const expectedProps = {
      show: true,
      onHide: mockOnHideLogoutModal,
      backgroundContentColor: 'Magenta',
    };

    expect(sideModal.props()).toEqual(expect.objectContaining(expectedProps));
  });

  it('should render SideModal with backgroundContentColor = Blue if userRole will be not AvatarByRole.REGULAR_EMPLOYEE', () => {
    const mockUseSelector = (selector: (state: unknown) => unknown) => {
      const state = { USER: { ...mockUserData, userRole: AvatarByRole.MEDICAL_DOCTOR } };
      return selector(state);
    };
    (useSelector as jest.Mock).mockImplementationOnce(mockUseSelector)
      .mockImplementationOnce(mockUseSelector);
    const wrapper = shallow(<LogoutModal />);
    const sideModal = wrapper.find(SideModal);
    const expectedProps = { backgroundContentColor: 'Blue' };

    expect(sideModal.props()).toEqual(expect.objectContaining(expectedProps));
  });

  it('should render img nested in Modal.Header with dulyLargeWhiteVioletIcon src props if userRole = AvatarByRole.REGULAR_EMPLOYEE', () => {
    const sideModal = wrapper.find(Modal.Header);
    const expectedProps = { children: <img alt="Large duly icon" src={dulyLargeWhiteVioletIcon} /> };

    expect(sideModal.props()).toEqual(expect.objectContaining(expectedProps));
  });

  it('should render img nested in Modal.Header with dulyLargeWhiteDarkBlueIcon src props if userRole != AvatarByRole.REGULAR_EMPLOYEE', () => {
    const mockUseSelector = (selector: (state: unknown) => unknown) => {
      const state = { USER: { ...mockUserData, userRole: AvatarByRole.MEDICAL_DOCTOR } };
      return selector(state);
    };
    (useSelector as jest.Mock).mockImplementationOnce(mockUseSelector)
      .mockImplementationOnce(mockUseSelector);
    const wrapper = shallow(<LogoutModal />);
    const sideModal = wrapper.find(Modal.Header);
    const expectedProps = { children: <img alt="Large duly icon" src={dulyLargeWhiteDarkBlueIcon} /> };

    expect(sideModal.props()).toEqual(expect.objectContaining(expectedProps));
  });

  it('should render Avatar with correct props', () => {
    const avatar = wrapper.find(Avatar);
    const {
      userRole,
      userData: { humanName: { familyName } },
    } = mockUserData;

    const expectedProps = {
      width: 12.31,
      src: 'mockSrc',
      alt: `mockFirstPrefixItem${familyName}`,
      role: userRole as AvatarByRole,
      hasBorder: true,
    };

    expect(avatar.props()).toEqual(expect.objectContaining(expectedProps));
  });

  it('should render element with correct human name', () => {
    const element = wrapper.find({ 'data-testid': 'human-name' });
    const familyName = mockUserData.userData.humanName.familyName;

    expect(element.text()).toEqual(`mockFirstPrefixItem${familyName}`);
  });

  it('should render element with only givenNames without prefixes if there are no prefixes', () => {
    const mockUseSelector = (selector: (state: unknown) => unknown) => {
      const state = {
        USER: {
          ...mockUserData,
          userData: {
            ...mockUserData.userData,
            humanName: {
              ...mockUserData.userData.humanName,
              prefixes: [],
            },
          },
        },
      };
      return selector(state);
    };
    (useSelector as jest.Mock).mockImplementationOnce(mockUseSelector)
      .mockImplementationOnce(mockUseSelector);
    const wrapper = shallow(<LogoutModal />);

    const element = wrapper.find({ 'data-testid': 'human-name' });
    const givenNames = mockUserData.userData.humanName.givenNames;

    expect(element.text()).toEqual(givenNames.join(' '));
  });
});
