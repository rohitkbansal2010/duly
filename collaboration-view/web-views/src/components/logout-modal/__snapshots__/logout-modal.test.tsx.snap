// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogoutModal should match snapshot 1`] = `
<SideModal
  backgroundContentColor="Magenta"
  onHide={[MockFunction]}
  show={true}
>
  <ModalHeader
    className="logoutModalHeader"
    closeButton={false}
    closeLabel="Close"
  >
    <img
      alt="Large duly icon"
      src="duly-large-white-violet.svg"
    />
  </ModalHeader>
  <ModalBody
    className="logoutModalBody"
  >
    <div
      className="logoutModalBodyPractitionerLogo"
    >
      <Avatar
        alt="mockFirstPrefixItemMorris"
        hasBorder={true}
        role="RegularEmployee"
        src="mockSrc"
        width={12.31}
      />
    </div>
    <div
      className="logoutModalBodyPractitionerHumanName"
      data-testid="human-name"
    >
      mockFirstPrefixItemMorris
    </div>
    <div
      className="logoutModalBodyQuestion"
    >
      Are you sure you want to log out?
    </div>
    <div>
      <button
        className="logoutModalBodyButtonLogout logoutModalBodyButtonViolet"
        data-testid="logout-button"
        onClick={[Function]}
      >
        Logout
      </button>
      <button
        className="logoutModalBodyButtonCancel logoutModalBodyButtonViolet"
        data-testid="cancel-button"
        onClick={[MockFunction]}
      >
        Cancel
      </button>
    </div>
  </ModalBody>
</SideModal>
`;
